core/src/main/scala/kafka/log/Log.scala

```diff
@@ -553,6 +553,17 @@ class Log(@volatile var logStartOffset: Long,
 
   /** Only used for ZK clusters when we update and start using topic IDs on existing topics */
   def assignTopicId(topicId: Uuid): Unit = {
+    // defensively check that any newly assign topic ID matches any that is already set
+    _topicId.foreach { current =>
+      if (!current.equals(topicId)) {
+        // we should never get here as the topic IDs should have been checked in becomeLeaderOrFollower
```

Comment:
nit: I think it's ok to leave this out. The point of adding the check is to reduce the coupling with ReplicaManager